atos/modules/MQTTBridge/src/mqttclient.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+#include "mqttclient.hpp"
+
+
+MQTTClient::MQTTClient(const std::string& host, const int& port, const std::string& username, const std::string& password) :
+	mqttHost(host),
+	mqttPort(port),
+	mqttUser(username),
+	mqttPass(password) {
+
+	// Each client needs a unique ID, adding a timestamp to the end of the ID should be enough
+	auto timeNanoseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
+	mqttClientID += "_" + std::to_string(timeNanoseconds);
+
+	mqttClient = std::make_unique<mqtt::async_client>(mqttHost + ":" + std::to_string(mqttPort), mqttClientID);
+	connectionOptions = mqtt::connect_options_builder()
+		.user_name(mqttUser)
+		.password(mqttPass)
+		.automatic_reconnect(std::chrono::seconds(MIN_RECONNECT_INTERVAL_SECONDS), std::chrono::seconds(MAX_RECONNECT_INTERVAL_SECONDS))
+		.clean_session(REMEMBER_RECONNECT_STATE)
+		.keep_alive_interval(std::chrono::seconds(KEEP_ALIVE_INTERVAL_SECONDS))
+		.finalize();
+}
+
+
+MQTTClient::~MQTTClient() {
+	if(mqttClient->is_connected()){
+        mqtt::disconnect_options disconnectOpts;
+        disconnectOpts.set_timeout(std::chrono::seconds(30));
+        mqttClient->disconnect(disconnectOpts)->wait();
+    }
+}
+
+
+/**
+ * @brief Connect to the MQTT broker with the connection options.
+ * 
+ */
+void MQTTClient::connect() {
+	mqttClient->connect(connectionOptions);
+}
+
+
+/**
+ * @brief Setup the MQTT subscription with the topics provided and connect to the broker.
+ * 
+ */
+void MQTTClient::setupSubscription(std::vector<std::string>& mqttTopics) {
+	mqttSubscriberCallback = std::make_unique<MQTTSubscriberCallback>(*mqttClient, connectionOptions, mqttTopics);
```

Comment:
Does this mean that a client can only subscribe to 1 topic at a time?

Reply:
The callback class creates a subscription for each topic in the vector when the client connects. I renamed the class to setupSubscription**s** to reflect this.